### PR TITLE
add support for recursing into arrays of structs

### PR DESCRIPTION
**Pull request by vincentcr**
currently the validator will not recurse into arrays of structs. For example, if I have:

```go
type X struct {
 A       int `validate:"nonzero"`
}

type Y struct {
 B int `validate:"nonzero"`
 Xs []X
}
```
the elements of `Xs` will not be validated. This pull request adds support for such a scenario.

### CODE REVIEW

**Comment by jameshartig** [comment]
@vincentcr there's a merge conflict with this can you resolve that?

File: validator_test.go [hunk 1]
```diff
@@ -266,6 +273,31 @@ func (ms *MySuite) TestBadParameter(c *C) {
...
+	tsErrs, ok := errs["Required"][0].(validator.ErrorMap)
+	c.Assert(ok, Equals, true)
+	c.Assert(tsErrs["A"], HasError, validator.ErrZeroValue)
+
+	s = TestStructSlice{E: 2, Required: []TestStruct{TestStruct{}, TestStruct{}}, Optional: []TestStruct{TestStruct{}}}
+	err = validator.Validate(s)
+	c.Assert(err, NotNil)
+	errs, ok = err.(validator.ErrorMap)
+	c.Assert(ok, Equals, true)
+	c.Assert(errs["Required"], HasLen, 2)
```
Comment (jameshartig) [inline validator_test.go:294]: Can you add a test here to make sure that `TestStructSlice{E: 3, Required: []TestStruct{}}` fails because the slice contains 0 elements? Also can you add a test `TestStructSlice{E: 3, Required: nil}` to make sure that errors too?

**Comment by jameshartig** [comment]
This should be fixed
